### PR TITLE
Sliders for storage sizes

### PR DESCRIPTION
**Pull request by mvollmer**
Such as partition and logical volumes.

### CODE REVIEW

**Comment by mvollmer** [comment]
@andreasn, please have a look.

screenshot from 2016-03-01 13-08-18

**Comment by andreasn** [comment]
Yo!

**Comment by andreasn** [comment]
I wonder if we need to cover the case of wanting to give something exactly 5200 MB. Right now you'll have to do a little dance where you first drag the slider to 700MB-ish and then change the number in the input. The alternative would be to have a dropdown with the different units, but maybe that gets messy with all the conventions it needs to do between the units. And also, what do we list in there? Everything from bytes to zettabytes?

**Comment by andreasn** [comment]
This is what GNOME Disks does.
The PB option just gives me silly numbers for this 4 GB (0,00000404 PB) usb stick, so if we do something similar, it would be good to filter that out somehow.

screenshot from 2016-03-01 12-53-15

**Comment by andreasn** [comment]
Quick mockup allowing to change dropdown value. Note that this assumes that the storage value don't change automatically.

mockup

**Comment by andreasn** [comment]
This needs a better error message:

screenshot from 2016-03-01 13-15-43

Could be "The size can only contain numbers"

**Comment by mvollmer** [comment]
@andreasn, please check this again.  Now there is a dropdown for the units.

**Comment by andreasn** [comment]
Nice!
Can you make the width of the entry 6em instead of 8em?
The size entry field feels a bit bit big right now. I played around a bit with the size and it seems setting it to 6em makes drives up to 100TB have plenty of space to convert to MB.

Apart from that I think it works really well!

**Comment by mvollmer** [comment]
> Can you make the width of the entry 6em instead of 8em?

Done.  I've also move the styles into storage.css.  Feel free to tweak them, of course!

**Comment by mvollmer** [comment]
About creating thin pool volumes.

When using lvcreate via storaged to create a thin pool, the given size is taken to be the size of the data volume of the pool.  lvcreate will also create some metadata volumes and they will take additional space.

This means that whacking the slider to the right when creating a thin pool will result in an error: the additional metadata volumes wont fit.

This should be fixed in storaged: the 'size' argument to CreateThinPoolVolume was always meant to be the total size that is taken out of the volume group, not the resulting size of the data subvolume of the pool.

https://github.com/storaged-project/storaged/pull/49

Should we wait until storaged is fixed and released?  Right now, I'd say yes.  (Hmm, we should have a test for this, actually.)

**Comment by andreasn** [comment]
> Should we wait until storaged is fixed and released? Right now, I'd say yes.

Yeah, I think I agree. I don't think this one is in any hurry.

**Comment by mvollmer** [comment]
>  (Hmm, we should have a test for this, actually.)

Added.

File: pkg/storaged/storage.css [hunk 1]
```diff
@@ -233,3 +233,87 @@
...
+    width: 18px;
+    height: 18px;
+    top: -6px;
+    right: -9px;
+    content: '.';
+    cursor: pointer;
+    border: 1px solid #267da1;
+    z-index: 1;
+}
+
```
Comment (andreasn) [inline pkg/storaged/storage.css]: I want to sync up the slide with the buttons a bit. Will create a branch for this.

**Comment by andreasn** [comment]
I added some style changes here https://github.com/andreasn/cockpit/commit/cb1c91969887bdaf04a9c77767cf9eb636c65ca0

Before:
screenshot from 2016-03-04 17-22-44

After:
screenshot from 2016-03-04 18-45-44

**Comment by andreasn** [comment]
This would fix issue #1116

**Comment by andreasn** [comment]
I've added a Patternfly issue about this and sent a mail to the patternfly list to see if anyone else is interested in a widget like this https://github.com/patternfly/patternfly/issues/210

**Comment by mvollmer** [comment]
The tests are failing as expected: We can't currently create a thin pool of maximum size.  This PR is blocked on the fix in storaged-project/storaged#49

**Comment by mulkieran** [comment]
Thanks, I am looking. I have thought a good deal about some parts of the problem of computing with and displaying address ranges in storage, here's my proof of work: https://github.com/mulkieran/justbytes.

This is, of course, a great improvement over what I was interacting with last week. But, just by looking at the mockups, it's a little hard for me to get every aspect of its use.

One thing I'll say, and this has been touched on before, contentiously, using XiB notation for IEC units removes ambiguity. Without it, whatever the convention may be, uncertainty about what value one is actually dealing with creeps into the naive user's mind. It happened to me on my Mac recently; I had no idea where to look to find out what the units actually were, and it was maddening. Whenever I'm doing the same kind of thing on a Linux machine I tend to know the details of the source code intimately, so it really doesn't count. From looking at the Cockpit interface, it seems like you tend to resist a whole lot of configuration, but would it make sense to add "Display Units" to the "Display Language" option?

I tend to think it would be nice if the units were displayed not using some kind of plain list, but instead that wheel widget, which includes +'s and -'s. This is because the units are not equal options, they have such an obvious implicit ordering. But, I have a sad feeling that that widget does not support discrete values in a list, you have to give it some kind of numeric range and then it fills in the gaps, somehow; at least that seems to be how it works in gtk widgets. So, it might not be possible without a lot of trouble.

I am very glad that you are making the choice of units by widget and _not_ allowing the user to enter the units as text. blivet/anaconda allows text input of units, and that makes the whole problem a lot more difficult.

andreas's suggestion that the units should be restricted to ones that make sense within the allowed range strikes me as a good idea. I think the bound would only ever need to be an upper bound. Maybe the pre-selected choice could always, reasonably, be the top one.

What happens when there is value in the numeric field and the user changes the units? My guess is that it seems less surprising for the total value to change...rather than for the total value to remain the same and the numeric value to be adjusted correspondingly.

I think the meaning of the slider might tend to disappear as the user adjusts values, I mean, you will lose track of what it's length represents. But it looks clean, and it does give you an idea of the proportion of the amount specified wrt. _something_. It is definitely better than no visual. Would it be possible to indicate the meaning of the whole length of the slider in some sort of symbolic way? Or by a numeric way where the display is identical to some other value on the screen, at least.

However, I'm not absolutely sure that it is worthwhile even making the slider interactive, just informational.
If the slider is interactive, then it might be desired to make it work in all cases. In some cases, the total range represented by the full range of the slider is very large, while the desired quantity might be small. In order to make it useful, it might seem necessary to make it logarithmic to give the user any real control through the slider. I think vtrefny, who works on blivet-gui told me that this problem had come up.

If the slider could be used to set the value, would the units dynamically change to accomodate the changing magnitude of the value. So, if you have a really big disk, and you want a small partition,
and you grab the slider and slide it way left, does the units dynamically drop to KiB or get you stuck in long fractionally quantities? Or is the precision limited, so you yourself have to change the units as you change the slider, so the value you're selecting does not get truncated to 0. This could have funny effects.

My experience is that there are a lot of situations where whatever number you specify ends up just being in the nature of a gentle hint or suggestion to the underlying tools. It seems like sometimes you want to specify something completely different than a precise value. I don't know how you would go about that.

Related, because users like to know what sizes things are when specifying new sizes...my justbytes library is able to indicate whether the number being shown is the same as the value it represents, or just an approximation to that value. Pretty soon it will be able to indicate the _relationship_ between the number displayed and the value it represents. I'm finding this useful in my own work, and I think it would be useful to others.

Developing justbytes also made me realize that the world would be a better place if sometime around the time I was born somebody had realized that using a base 1024 number system would have been handy in this case and pretty easy to adjust to. In base 1024, 931,526 MiB is 931.526 GiB (931 GiB and 526 MiB),
which is not hard to follow. In base 10, same values, 953870 MiB is 931.513671875 GiB, which is hardly
so clear. Personally, I would love it if you could toggle between base 10 and 1024 number systems just by pressing the value itself.

File: pkg/storaged/dialog.js [hunk 2]
```diff
@@ -72,16 +76,94 @@ define([
...
+        function size_text_changed(event) {
+            var input = $(this);
+            var parent = input.parents('.size-slider');
+            var unit = parent.find('.size-unit');
+            var unit_val = +unit.val();
+            var slider = parent.find('.slider');
+            var max = parent.data('max');
+            var value = +input.val() * unit_val;
+
+            // As a special case, if the user types something that
```
Comment (mulkieran) [inline pkg/storaged/dialog.js]: This seems like a bad idea to me...I've seen the general approach taken also in blivet/anaconda and it made things _very_ hard.
I think it would be a better design, to:
* Mark all approximate values as exactly that, approximate.
* For all displayed values, allow the user to click on them to see an alternate display in whole bytes, at least, or perhaps, the smallest units that allows no fractional quantities (justbytes is able to calculate this, but it is a library, not a GUI).
* If the user wants exactly that amount, let them choose exactly those units and put in exactly that number.

I see that the slider represents a fraction between 0 and 1. Maybe it would be possible to make it especially obvious if the value was exactly 1, graphically.

File: pkg/storaged/dialog.js [hunk 3]
```diff
@@ -72,16 +76,94 @@ define([
...
+        }
+
+        function size_slider_changed(event, value) {
+            var parent = $(this).parents('.size-slider');
+            var input = parent.find('.size-text');
+            var unit = parent.find('.size-unit');
+            var max = parent.data('max');
+            var round = parent.data('round');
+
+            value *= max;
```
Comment (mulkieran) [inline pkg/storaged/dialog.js]: Rounding is another tricky bit. See the justbytes roundTo() method which is a bit more complicated than your approach because of problems that I have observed arising. https://github.com/mulkieran/justbytes/blob/develop-0.05/src/justbytes/_size.py#L516. One important difference is that the rounding method is specified, the other that lower and upper bounds are specified.

**Comment by mvollmer** [comment]
> Thanks, I am looking. 

Thanks a lot for the feedback!

**Comment by mvollmer** [comment]
> One thing I'll say, and this has been touched on before, contentiously, using XiB notation for IEC units removes ambiguity.

Yes, I agree.  We have been idle, waiting for sound advice, but I think we now need to sort this out.  I will make a separate PR that makes us use the correct unit names.

File: pkg/storaged/dialog.js (hunk 3 above)
Comment (mvollmer) [inline pkg/storaged/dialog.js]: Yes, we need to at least make sure that using the slider never results in a value less than 0 or greater than max, due to rounding errors.

File: pkg/storaged/dialog.js (hunk 2 above)
Comment (mvollmer) [inline pkg/storaged/dialog.js]: How about adding "Bytes" as a unit?  That way, the user can enter a precise and unambiguous number.

Comment (mvollmer) [inline pkg/storaged/dialog.js]: If we add "Bytes", I guess we also need to add "KiB"....

**Comment by mvollmer** [comment]
> From looking at the Cockpit interface, it seems like you tend to resist a whole lot of configuration, but would it make sense to add "Display Units" to the "Display Language" option?

@andreasn?

**Comment by mvollmer** [comment]
> What happens when there is value in the numeric field and the user changes the units? My guess is that it seems less surprising for the total value to change...rather than for the total value to remain the same and the numeric value to be adjusted correspondingly.

Right now, the total value remains the same, and the text in the input field changes.  @andreasn, what's your opinion?

**Comment by mvollmer** [comment]
> If the slider could be used to set the value, would the units dynamically change to accomodate the changing magnitude of the value.

Right now, the units never change automatically.  The default is selected when the dialog opens, but while the dialog is open, they only change when the user does it explicitly.

**Comment by mvollmer** [comment]
> Or is the precision limited, so you yourself have to change the units as you change the slider, so the value you're selecting does not get truncated to 0.

Precision is limited to one digit after the comma, but the text remains at "0.1" unless the value is exactly zero.

**Comment by mvollmer** [comment]
> My experience is that there are a lot of situations where whatever number you specify ends up just being in the nature of a gentle hint or suggestion to the underlying tools. 

This is true.  Storaged/UDisks2 round everything to megabytes, and so does Cockpit.  For LVM2, we round to the extent size, which is usually larger than 1 MiB, I think.

**Comment by andreasn** [comment]
> From looking at the Cockpit interface, it seems like you tend to resist a whole lot of configuration, but would it make sense to add "Display Units" to the "Display Language" option?

MB/GB/TB was what was recommended to us from the Red Hat UX team.
I would not like to see a change of units in this pull req, but rather as a separate issue. And with involvement from Patternfly people.

**Comment by andreasn** [comment]
> If the slider could be used to set the value, would the units dynamically change to accomodate the changing magnitude of the value.

The first version of the dialog worked like this, but it felt very jarring. This way you can set the unit you want (say GB) and remain certain that a control won't change units under your feet.

**Comment by mvollmer** [comment]
> I added some style changes here andreasn@cb1c919

Imported to this pull request.

File: pkg/storaged/dialog.js (hunk 3 above)
Comment (mvollmer) [inline pkg/storaged/dialog.js]: Done.

**Comment by mvollmer** [comment]
> Storaged/UDisks2 round everything to megabytes, and so does Cockpit.

This is actually false. Storaged/UDisks2 rounds the starting offset to the nearest MiB, but the size is not rounded.  So I'll remove this rounding from Cockpit also.

**Comment by mulkieran** [comment]
Great!

I'm not sure what rationale storaged uses for rounding offsets
to nearest mebibyte, and I'm not sure what the impact would be,
so overall this definitely feels like a step in the right direction.

----- Original Message -----

**Comment by mulkieran** [comment]
My suggestion really isn't about changing units from IEC to SI; its
about un-confusing the user. I myself have a strong preference for IEC,
principally because these units reflect the actual structure of memory addressing,
and are therefore natural for it. This, in spite of the awkwardness of
representation in a base-10 number system.

I talked to Ju Lim a few weeks ago and she told me that it was
somewhat due to her instigation that the notation was changed and that
it caused a lot of contention at the time. She tells me that it has
been sort of a Linux standard to use XiB, but that among our competitors
XB notation is always used. I don't know anything about the industry standard,
so I can not say. She also said that a user can infer that when memory is being
sold XB means SI units, but when it is being divided into partitions XB means IEC
units. That is not true of my own experience, and so I think it isn't
true of others. Also, she gave the very strong impression that she would
not want to take the lead on this issue.

So, I'm looking for a way that the naive Cockpit (or any) user can know what it means
when they see 1.52 GB. Right now, the number of possibilities is very large
due to a combination of rounding and uncertainty about what the GB means.
It is to be hoped that at least, due to the uniformity that Ric Wheeler is
always pushing for, that Red Hat project will become consistent, and anybody
using a Red Hat project not about sales will know that GB is meant to indicate
IEC units. But in the meantime, which could easily be 10 years, I'm wondering
if there is a more direct way to reassure the user that yes, that GB means IEC
units and the number of possible values that 1.52 GB actually represents has
just been halved, and is all clustered around one set of values.
Having them explicitly make a choice, which defaults to IEC, is one way to do
this. Hence this particular proposal.

I'll open a pattern fly issue about disambiguation of the meaning of these
overloaded symbols and numbers.

- mulhern

----- Original Message -----

**Comment by andreasn** [comment]
> I'll open a pattern fly issue about disambiguation of the meaning of these
overloaded symbols and numbers.

Thanks! Would be good to clear this up.

File: pkg/base1/cockpit.js [hunk 4]
```diff
@@ -2248,6 +2248,22 @@ function full_scope(cockpit, $, po) {
...
+        /* non-zero values should never appear zero */
+        if (number > 0 && number < 0.1)
+            number = 0.1;
+        else if (number < 0 && number > -0.1)
+            number = -0.1;
+
+        /* TODO: Make the decimal separator translatable */
+
+        /* only show as integer if we have a natural number */
+        if (number % 1 === 0)
```
Comment (mulkieran) [inline pkg/base1/cockpit.js]: Probably use ```isInteger()``` here?

**Comment by mulkieran** [comment]
For the record, I think a general solution that indicates that
a value is approximate and what relation it has to the actual value,
is a better solution than this special
casing of a particular value, which could mislead by orders of
magnitude.

But I understand that you don't have the liberty to innovate
to that degree.

I'll take this issue up on pattern fly and wherever else I can manage.

- mulhern

----- Original Message -----

**Comment by mulkieran** [comment]
Done: https://github.com/patternfly/patternfly/issues/216.

**Comment by mvollmer** [comment]
This is unblocked on fedora-24.

**Comment by mvollmer** [comment]
This is unblocked on rhel-7 as well, via a known issue.

File: pkg/base1/cockpit.js (hunk 4 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: This is old code, actually, and we know that it works.  What's the advantage of changing it?

**Comment by mvollmer** [comment]
I think this is ready to go in.  The test failures seem unrelated.

**Comment by mvollmer** [comment]
> For the record, I think a general solution that indicates that a value is approximate and what relation it has to the actual value, is a better solution than this special casing of a particular value, which could mislead by orders of magnitude.

I can't see how this can happen.  Can you give a concrete example?

**Comment by stefwalter** [comment]
This needs a rebase

File: pkg/base1/cockpit.js (hunk 4 above)
Comment (mulkieran) [inline pkg/base1/cockpit.js]: I suspect isInteger() is old too, and I sure hope it works :) But to spend time actually discussing a change so minuscule...I'm confident that we both have more important things to do.

**Comment by mulkieran** [comment]
Regarding the adjustment that occurs in the ```format_number()``` function...

I'm assuming that this adjustment is necessary...that is that you might have a number that is less than 0.1 which you wish to display, for example. The number would be displayed as 0.1 and would probably be interpreted as, at least approximately, 0.1. If the units are TiB and the actual value is 2 KiB, then the interpretation of the value by the user is several orders of magnitude off. Whether or not this situation could actually occur is dependent on the context in which the function is invoked, which is a hard thing to constrain or reason about.

At this point, my justbytes module, now available in fedora updates at least, actually provides an indication of the relationship of the displayed value to the one it represents, and I do feel that this is non-deceptive and helpful info to the user.

```
>>> import justbytes
>>> from justbytes import *
>>> big_num = Size(355, TiB)
>>> small_num = Size(2, KiB)
>>> tricky = big_num + small_num
>>> str(tricky)
'>355.00 TiB'
>>> str(big_num)
'355 TiB'
>>> str(small_num)
'2 KiB'
>>> tricky.getStringInfo(SizeConfig.STR_CONFIG)
(Radix(True,[3, 5, 5],[0, 0],[],10), -1, TiB)
```
The result above is a 3-tuple indicating the value to display, the relation of that value to the actual value, and the units. That info is what is used to actually come up with the string value.

Note that base-1024 gives the clearest information:
```
>>> tricky.getString(StrConfig(base=1024, max_places=None), SizeConfig.DISPLAY_CONFIG)
'355.0,0,2 TiB'
```
355 TiB, 0 GiB, 0 MiB, 2 KiB, but I know that that is an uphill battle.

The largest units that allow a non-fractional quantity, while helpful in some situations, like specifying a value precisely, is certainly not informative to the typical user:

```
>>> tricky.getString(StrConfig(exact_value=True), SizeConfig.DISPLAY_CONFIG)
'381178347522 KiB'
```

**Comment by mvollmer** [comment]
> Regarding the adjustment that occurs in the format_number() function...

I see.  I thought you were talking about the adjustment in `size_text_changed`.

I see your point.  Should we pursue this outside of this PR, though?

File: pkg/base1/cockpit.js [hunk 5]
```diff
@@ -992,6 +992,22 @@ function basic_scope(cockpit, jquery) {
         return fmt.replace(fmt_re, function(m, x, y) { return args[x || y] || ""; });
     };
 
+    cockpit.format_number = function format_number(number) {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: If we're adding public stable API it should be documented and tested.

File: pkg/base1/cockpit.js [hunk 6]
```diff
@@ -1068,6 +1071,31 @@ function basic_scope(cockpit, jquery) {
         return format_units(number, byte_suffixes, factor, separate);
     };
 
+    cockpit.get_byte_units = function get_byte_units(guide_value, factor) {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: This needs documentation. How about calling it 'parse_byte_units' to match with the 'format_bytes' naming?

File: pkg/storaged/index.html [hunk 7]
```diff
@@ -762,6 +762,29 @@ <h4 class="modal-title">{{Title}}</h4>
...
+                      <table>
+                        <tr>
+                          <td>
+                            <div class="slider"></div>
+                          </td>
+                          <td>
+                            <input class="form-control size-text">
+                          </td>
+                          <td>
+                            <select class="form-control selectpicker size-unit">
```
Comment (stefwalter) [inline pkg/storaged/index.html]: Would it be possible to avoid additional use of bootstrap-select, especially in templates? It's something we're trying to get rid of. playground/patterns.html includes an example of how to avoid it. Or is that too hard to do here?

File: pkg/base1/cockpit.js (hunk 6 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: The function does not do any parsing, it returns a list of units to use in a UI dropdown.

File: pkg/base1/cockpit.js (hunk 5 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Indeed.  I had somehow misssed that `format_bytes` et al are public API...

File: pkg/base1/cockpit.js (hunk 6 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Documented and tested.

File: pkg/storaged/index.html (hunk 7 above)
Comment (mvollmer) [inline pkg/storaged/index.html]: Very good suggestion.  I'll try to remove bootstrap-select from all of storage.

Comment (mvollmer) [inline pkg/storaged/index.html]: Done.

**Comment by mvollmer** [comment]
> But in the meantime, which could easily be 10 years,

Soon... :-)
